moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.cpp

```diff
@@ -0,0 +1,278 @@
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author    : Adam Pettinger
+   Desc      : Declares, loads, and checks ServoParameters for Servo
+   Title     : servo_parameters.cpp
+   Project   : moveit_servo
+   Created   : 07/02/2020
+*/
+
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <type_traits>
+
+#include <moveit_servo/servo_parameters.h>
+
+namespace moveit_servo
+{
+template <typename T>
+bool declareAndGetParam(const std::string& param_name, T& output_value, rclcpp::Node::SharedPtr& node,
+                        const rclcpp::Logger& logger)
+{
+  bool got_param = false;
+  try
+  {
+    node->declare_parameter<T>(param_name, T{});
+    got_param = node->get_parameter(param_name, output_value);
+  }
+  catch(const rclcpp::exceptions::InvalidParameterTypeException& e)
+  {
+    // Catch a <double> parameter written in the yaml as "1" being considered an <int>
+    if(std::is_same<T, double>::value)
+    {
+      int value;
+      node->undeclare_parameter(param_name);
+      node->declare_parameter<int>(param_name, 0);
+      got_param = node->get_parameter(param_name, value);
+      output_value = value;
+    }
+    else
+    {
+      throw e;
+    }
+  }
+
+  if (got_param)
+  {
+    return true;
+  }
+  else
+  {
+    RCLCPP_WARN_STREAM(logger, "Unable to get parameter: \'" << param_name << "\'. Please check YAML file");
+    return false;
+  }
+}
+
+/**
+ * Declares, reads, and validates parameters used for moveit_servo
+ * @param node Shared ptr to node that will the parameters will be declared in. Params should be defined in launch/config files
+ * @param logger Logger for outputting warnings about the parameters
+ * @param parameters The set up parameters that will be updated. After this call, they can be used to start a Servo instance
+ * @param ns Parameter namespace (as loaded in launch files). Defaults to "moveit_servo" but can be changed to allow multiple arms/instances
+ * @return true if all parameters were loaded and verified successfully, false otherwise
+ */
+bool readParameters(rclcpp::Node::SharedPtr& node, const rclcpp::Logger& logger, ServoParametersPtr& parameters, std::string ns = "moveit_servo")
```

Comment:
`const shared_ptr<rclcpp::Node>&` here too.

Looking at this one I'm reminded why I like the output parameter (parameters) to be the first parameter.  You can't make it the last one as it would conflict with default parameters which have to go at the end.  Also, when you make it the first parameter the output parameter it is more intuitive as the left of the operator= is the output:

output = function(inputs)
function(output, inputs)

Reply:
Moved it to the start on both